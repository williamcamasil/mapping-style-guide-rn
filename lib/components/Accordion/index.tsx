import React, {
  useCallback,
  useRef,
  useState,
} from 'react';
import {
  Animated,
  StyleProp,
  LayoutChangeEvent,
  View,
  ViewStyle,
  StyleSheet,
  TouchableWithoutFeedback,
} from 'react-native';

import { useDidMountAndUpdate } from 'mapping-context-rn';

import { useTextStyles, useViewStyles } from '../../hooks/useStyles';
import { withTheme } from '../../theme';
import { AppThemeType } from '../../theme/default';
import CollapseArrow from '../Collapse/CollapseArrow';
import Divider from '../Divider';
import Spacer from '../Spacer';
import Text from '../Text';

export type AccordionPropsType = {
  theme: AppThemeType;
  title: string;
  description: string;
  testID?: string;
  expanded?: boolean;
  style?: StyleProp<ViewStyle>;
  onPress?: () => void;
  onCollapseEnd?: () => void;
};

const styles = StyleSheet.create({
  divider: {
    position: 'absolute',
    bottom: 0,
    left: 0,
    right: 0,
  },
  content: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
  },
  headerWrapper: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    flex: 1,
    alignItems: 'center',
  },
});

const Accordion = ({
  theme,
  testID,
  expanded,
  title,
  description,
  style,
  onPress,
  onCollapseEnd,
  ...others
}: AccordionPropsType) => {
  const [isExpanded, setIsExpanded] = useState(Boolean(expanded));
  const animationHeight = useRef(new Animated.Value(0));
  const [headerHeight, setHeaderHeight] = useState(0);
  const [contentHeight, setContentHeight] = useState(0);

  useDidMountAndUpdate(() => {
    setIsExpanded(Boolean(expanded));
  }, [expanded]);

  const handleContentLayout = useCallback((event: LayoutChangeEvent) => {
    setContentHeight(event.nativeEvent.layout.height);
  }, []);

  const handleHeaderLayout = useCallback((event: LayoutChangeEvent) => {
    const height = event.nativeEvent.layout.height + theme.spacings.lSmall;
    setHeaderHeight(height);
    animationHeight.current.setValue(height);
  }, [theme.spacings.lSmall]);

  useDidMountAndUpdate(() => {
    const newHeight = isExpanded ? contentHeight : headerHeight;

    if (!newHeight) return undefined;

    const animation = Animated.parallel([
      Animated.spring(animationHeight.current, {
        toValue: newHeight,
        useNativeDriver: false,
      }),
    ]);

    animation.start(() => {
      onCollapseEnd?.();
    });

    return () => {
      animation.stop();
    };
  }, [isExpanded, contentHeight, headerHeight]);

  const handleCollapse = useCallback(() => {
    onPress?.();
    setIsExpanded(current => !current);
  }, [onPress]);

  const containerStyles = useViewStyles(() => [
    {
      overflow: 'hidden',
      height: contentHeight ? animationHeight.current : undefined,
    },
    style,
  ], [contentHeight, style]);

  const headerTouchableStyles = useViewStyles(() => [
    {
      paddingVertical: theme.spacings.sMedium,
      paddingHorizontal: theme.spacings.sXXS,
      paddingBottom: theme.spacings.sZero,
    },
  ], [theme]);

  const titleStyles = useTextStyles(() => [
    {
      marginRight: theme.spacings.sQuark,
    },
  ], [theme.spacings.sQuark]);

  const renderDescription = () => {
    const numberOfLines = isExpanded ? undefined : 1;
    return (
      <TouchableWithoutFeedback onPress={handleCollapse}>
        <View>
          <Spacer size={theme.spacings.sNano} />
          <Text testID="accordion-description" numberOfLines={numberOfLines}>{description}</Text>
          <Spacer size={theme.spacings.sLarge} />
        </View>
      </TouchableWithoutFeedback>
    );
  };

  return (
    <Animated.View testID={testID} style={containerStyles}>
      <View testID="accordion-content" style={styles.content} onLayout={handleContentLayout}>
        <TouchableWithoutFeedback
          {...others}
          testID="accordion-touchable"
          onLayout={handleHeaderLayout}
          onPress={handleCollapse}
          style={headerTouchableStyles}
        >
          <View>
            <View style={styles.headerWrapper}>
              <Text
                testID="accordion-title"
                weight="bold"
                style={titleStyles}
                variant="body"
                color="neutralGray700"
              >
                {title}
              </Text>
              <CollapseArrow expanded={isExpanded} />
            </View>
          </View>
        </TouchableWithoutFeedback>
        {renderDescription()}
      </View>
      <Divider style={styles.divider} />
    </Animated.View>
  );
};

export default withTheme(Accordion);
