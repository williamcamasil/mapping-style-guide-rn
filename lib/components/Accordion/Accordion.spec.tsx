import React from 'react';

import {
  fireEvent, render, waitFor,
} from '@testing-library/react-native';

import Accordion, { AccordionPropsType } from '.';
import { DefaultTheme } from '../../theme';

const fakeHeaderLayoutEvent = {
  nativeEvent: {
    layout: {
      height: 50,
    },
  },
};

const fakeContentLayoutEvent = {
  nativeEvent: {
    layout: {
      height: 200,
    },
  },
};

const renderAccordion = async (props: Omit<AccordionPropsType, 'theme'>) => {
  const handleCollapseEnd = jest.fn();

  const result = render(
    <Accordion
      {...props}
      onCollapseEnd={() => {
        handleCollapseEnd();
        props.onCollapseEnd?.();
      }}
    />,
  );

  const accordionContent = result.getByTestId('accordion-content');
  const accordionTouchable = result.getByTestId('accordion-touchable');

  fireEvent(accordionTouchable, 'layout', fakeHeaderLayoutEvent);
  fireEvent(accordionContent, 'layout', fakeContentLayoutEvent);

  await waitFor(() => {
    expect(handleCollapseEnd).toBeCalled();
  }, { timeout: 2000 });

  return result;
};

describe('Accordion snapshots', () => {
  it('Should match default snapshot', async () => {
    const result = await renderAccordion({
      title: 'My title',
      description: 'My description',
    });
    expect(result.toJSON()).toMatchSnapshot();
  });

  it('Should match snapshot when expanded', async () => {
    const result = await renderAccordion({
      title: 'My title',
      description: 'My description',
      expanded: true,
    });

    expect(result.toJSON()).toMatchSnapshot();
  });
});

describe('Accordion events', () => {
  it('Should have the correct height initially', async () => {
    const result = await renderAccordion({
      testID: 'accordion',
      title: 'My title',
      description: 'My extensive description to change the height of the accordion and test the animation.',
    });

    const accordion = result.getByTestId('accordion').props.style;

    await waitFor(() => {
      expect(accordion.height).toBe(
        fakeHeaderLayoutEvent.nativeEvent.layout.height + DefaultTheme.spacings.lSmall,
      );
    });
  });

  it('Should have the correct height after expanded', async () => {
    const result = await renderAccordion({
      testID: 'accordion',
      title: 'My title',
      description: 'My extensive description to change the height of the accordion and test the animation.',
      expanded: true,
    });

    const accordion = result.getByTestId('accordion').props.style;

    await waitFor(() => {
      expect(accordion.height).toBe(
        fakeContentLayoutEvent.nativeEvent.layout.height,
      );
    });
  });

  it('Should call onCollapseEnd after animation is over', async () => {
    const onPressMock = jest.fn();
    const onCollapseEndMock = jest.fn();

    const result = await renderAccordion({
      testID: 'accordion-component',
      onPress: onPressMock,
      title: 'Title Mock',
      description: 'My extensive description to change the height of the accordion and test the animation.',
      onCollapseEnd: onCollapseEndMock,
    });

    const accordionTouchable = result.getByTestId('accordion-touchable');
    fireEvent.press(accordionTouchable);

    expect(onPressMock).toBeCalled();

    await waitFor(() => {
      expect(onCollapseEndMock).toBeCalled();
    });
  });

  it('Should have its description with numberOfLines equals to 1 when collapsed', async () => {
    const onPressMock = jest.fn();
    const onCollapseEndMock = jest.fn();

    const result = await renderAccordion({
      testID: 'accordion-component',
      onPress: onPressMock,
      title: 'Title Mock',
      description: 'My extensive description to change the height of the accordion and test the animation.',
      onCollapseEnd: onCollapseEndMock,
      expanded: false,
    });

    const accordionDescription = result.getByTestId('accordion-description');

    expect(accordionDescription.props).toHaveProperty('numberOfLines', 1);
  });

  it('Should have its description with numberOfLines as undefined when expanded', async () => {
    const onPressMock = jest.fn();
    const onCollapseEndMock = jest.fn();

    const result = await renderAccordion({
      testID: 'accordion-component',
      onPress: onPressMock,
      title: 'Title Mock',
      description: 'My extensive description to change the height of the accordion and test the animation.',
      onCollapseEnd: onCollapseEndMock,
      expanded: true,
    });

    const accordionDescription = result.getByTestId('accordion-description');

    expect(accordionDescription.props).not.toHaveProperty('numberOfLines');
  });
});
