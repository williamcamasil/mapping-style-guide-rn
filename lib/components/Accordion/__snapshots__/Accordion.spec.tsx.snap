// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion snapshots Should match default snapshot 1`] = `
<View
  collapsable={false}
  style={
    {
      "height": 106,
      "overflow": "hidden",
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      {
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
    testID="accordion-content"
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onLayout={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="accordion-touchable"
    >
      <View
        style={
          {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <Text
          allowFontScaling={false}
          style={
            {
              "color": "#3E4149",
              "flexShrink": 1,
              "fontFamily": "Averta CY",
              "fontSize": 16,
              "fontWeight": "bold",
              "letterSpacing": -0.02,
              "lineHeight": 22,
              "marginRight": 4,
            }
          }
          testID="accordion-title"
        >
          My title
        </Text>
        <RNSVGSvgView
          align="xMidYMid"
          bbHeight="16"
          bbWidth="16"
          fill="none"
          focusable={false}
          height={16}
          meetOrSlice={0}
          minX={0}
          minY={0}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderWidth": 0,
              },
              {
                "flex": 0,
                "height": 16,
                "width": 16,
              },
            ]
          }
          vbHeight={16}
          vbWidth={16}
          width={16}
        >
          <RNSVGGroup
            fill={null}
            propList={
              [
                "fill",
              ]
            }
          >
            <RNSVGPath
              d="m4 6 4 4 4-4"
              fill={
                {
                  "payload": 4278190080,
                  "type": 0,
                }
              }
              propList={
                [
                  "stroke",
                  "strokeWidth",
                  "strokeLinecap",
                  "strokeLinejoin",
                ]
              }
              stroke={
                {
                  "payload": 4290954697,
                  "type": 0,
                }
              }
              strokeLinecap={1}
              strokeLinejoin={1}
              strokeWidth="2"
            />
          </RNSVGGroup>
        </RNSVGSvgView>
      </View>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <Text
        style={
          {
            "height": 8,
            "width": 8,
          }
        }
      />
      <Text
        allowFontScaling={false}
        numberOfLines={1}
        style={
          {
            "color": "#676D79",
            "flexShrink": 1,
            "fontFamily": "Averta CY",
            "fontSize": 16,
            "fontWeight": "normal",
            "letterSpacing": -0.02,
            "lineHeight": 22,
          }
        }
        testID="accordion-description"
      >
        My description
      </Text>
      <Text
        style={
          {
            "height": 28,
            "width": 28,
          }
        }
      />
    </View>
  </View>
  <View
    style={
      {
        "backgroundColor": "#F0F0F2",
        "bottom": 0,
        "height": 1,
        "left": 0,
        "marginBottom": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "position": "absolute",
        "right": 0,
      }
    }
  />
</View>
`;

exports[`Accordion snapshots Should match snapshot when expanded 1`] = `
<View
  collapsable={false}
  style={
    {
      "height": 200,
      "overflow": "hidden",
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      {
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
    testID="accordion-content"
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onLayout={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="accordion-touchable"
    >
      <View
        style={
          {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <Text
          allowFontScaling={false}
          style={
            {
              "color": "#3E4149",
              "flexShrink": 1,
              "fontFamily": "Averta CY",
              "fontSize": 16,
              "fontWeight": "bold",
              "letterSpacing": -0.02,
              "lineHeight": 22,
              "marginRight": 4,
            }
          }
          testID="accordion-title"
        >
          My title
        </Text>
        <RNSVGSvgView
          align="xMidYMid"
          bbHeight="16"
          bbWidth="16"
          fill="none"
          focusable={false}
          height={16}
          meetOrSlice={0}
          minX={0}
          minY={0}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderWidth": 0,
              },
              {
                "flex": 0,
                "height": 16,
                "width": 16,
              },
            ]
          }
          vbHeight={16}
          vbWidth={16}
          width={16}
        >
          <RNSVGGroup
            fill={null}
            propList={
              [
                "fill",
              ]
            }
          >
            <RNSVGPath
              d="M12 10 8 6l-4 4"
              fill={
                {
                  "payload": 4278190080,
                  "type": 0,
                }
              }
              propList={
                [
                  "stroke",
                  "strokeWidth",
                  "strokeLinecap",
                  "strokeLinejoin",
                ]
              }
              stroke={
                {
                  "payload": 4290954697,
                  "type": 0,
                }
              }
              strokeLinecap={1}
              strokeLinejoin={1}
              strokeWidth="2"
            />
          </RNSVGGroup>
        </RNSVGSvgView>
      </View>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <Text
        style={
          {
            "height": 8,
            "width": 8,
          }
        }
      />
      <Text
        allowFontScaling={false}
        style={
          {
            "color": "#676D79",
            "flexShrink": 1,
            "fontFamily": "Averta CY",
            "fontSize": 16,
            "fontWeight": "normal",
            "letterSpacing": -0.02,
            "lineHeight": 22,
          }
        }
        testID="accordion-description"
      >
        My description
      </Text>
      <Text
        style={
          {
            "height": 28,
            "width": 28,
          }
        }
      />
    </View>
  </View>
  <View
    style={
      {
        "backgroundColor": "#F0F0F2",
        "bottom": 0,
        "height": 1,
        "left": 0,
        "marginBottom": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "position": "absolute",
        "right": 0,
      }
    }
  />
</View>
`;
